apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1092 @@
+of the finer implications of that, such as how it affects the creation
+of universal lockfiles, that attempt to account for all valid environment
+marker combinations.
+
+## Let's talk about security
+
+The most important part of the design that we're still working on is security.
+The installer, including all its (possibly vendored) dependencies, is sometimes
+found operating with elevated privileges, for example when it is used to install
+packages for all users, on a multi-user system. Variant provider plugins will
+be operating with the same level of privileges, and this has significant
+implications on how they are developed.
+
+<div style={{ textAlign: "center" }}>
+<figure style={{width: 'auto', margin: '0 2em', display: 'inline-block', verticalAlign: 'top'}}>
+  <img src="/posts/python-wheels-from-tags-to-variants/security-scopes.png"
+    width="906" height="458" alt="Change in scopes of privileges due to variant
+plugins. Prior to their support, only installer (and its dependencies) would be
+executed at package install (possibly with elevated privileges), and wheel
+contents would only be executed when the package/application is used.
+After the change, also variant plugin (and its dependencies) is executed
+at package install, which means executing contents of the downloaded wheels." />
+  <figcaption>Fig. 10. Scopes of privileges when executing code</figcaption>
+</figure>
+</div>
+
+There are two main risks being considered here. First, malicious code could
+be introduced into an existing variant provider or one of its dependencies;
+this is a similar risk to how such code could be introduced in the installer
+itself. Second, a new provider could be created that includes malicious code
+by design, and existing packages could start using it.
+
+It can be noted that these risks are similar to Python package build backends,
+that are also downloaded dynamically during the install process and that execute
+arbitrary code throughout the build. However, this risk does not apply
+to installing from wheels, and can be mitigated in environment with higher
+security expectations by disabling source builds, e.g. via `pip install --only-binary :all:`.
+
+Similarly, there are multiple ways in which the risks from using variant wheels
+could be mitigated.
+I have already mentioned the possibility of using frozen variant provider output
+instead of running the plugin locally. However, this requires additional effort
+from the user and will primarily  be used in setups with high security
+requirements.
+
+To avoid indirect supply chain attacks, variant providers could recursively pin
+or vendor their dependencies. The latter option is especially interesting, since
+it can be easily enforced at installer level — by forcing provider plugins
+to be installed without dependencies. This way, we mitigate the risk of a new
+version of a compromised dependency being immediately deployed to end users.
```

Comment:
Should this also say that we expect most providers to be small and have few or no dependencies? E.g., `nvidia-provider-plugin` depends only on `packaging`, and vendors it in the same way that `pip` does.

Reply:
Added.